plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php

```diff
@@ -0,0 +1,215 @@
+		$this->store_entity_version( $entity_type, $entity_id, $version, true );
+		return $version;
+	}
+
+	/**
+	 * Store the entity version in a transient with a filterable TTL.
+	 *
+	 * @param string     $entity_type Entity type.
+	 * @param string|int $entity_id   Entity ID.
+	 * @param string     $version     The version to store.
+	 * @param bool       $is_new      Whether this is a new version (true) or a refresh (false).
+	 * @return bool True on success, false on failure.
+	 */
+	protected function store_entity_version( string $entity_type, $entity_id, string $version, bool $is_new ): bool {
+		$transient_name = "wc_entity_version_{$entity_type}_{$entity_id}";
+
+		/**
+		 * Filter the TTL for entity version cache.
+		 *
+		 * @param int        $ttl         Time to live in seconds. Default 1 day.
+		 * @param string     $entity_type The type of the entity.
+		 * @param string|int $entity_id   The ID of the entity.
+		 *
+		 * @since 10.4.0
+		 */
+		$ttl = apply_filters( 'woocommerce_cached_entity_version_ttl', DAY_IN_SECONDS, $entity_type, $entity_id );
+		$ttl = max( 0, (int) $ttl );
+
+		$result = $this->set_cached( $transient_name, $version, $ttl );
+
+		/**
+		 * Fires after an entity version has been generated or modified.
+		 *
+		 * @since 10.4.0
+		 *
+		 * @param string     $entity_type The type of the entity.
+		 * @param string|int $entity_id   The ID of the entity.
+		 * @param int        $ttl         Time to live in seconds.
+		 * @param bool       $is_new      Whether this is a new version (true) or a refresh of existing version (false).
+		 */
+		do_action( 'woocommerce_entity_version_cached', $entity_type, $entity_id, $ttl, $is_new );
+
+		return $result;
+	}
+
+	/**
+	 * Forget the version of an entity by deleting its transient.
+	 *
+	 * @param string     $entity_type Entity type.
+	 * @param string|int $entity_id   Entity ID.
+	 * @return bool True on success, false on failure.
+	 * @throws \InvalidArgumentException If entity_type or entity_id are invalid.
+	 */
+	public function forget_entity_version( string $entity_type, $entity_id ): bool {
+		$this->validate_input( $entity_type, $entity_id );
+
+		$transient_name = "wc_entity_version_{$entity_type}_{$entity_id}";
+		$result         = $this->delete_cached( $transient_name );
+
+		/**
+		 * Fires after an entity version has been explicitly deleted.
+		 *
+		 * @since 10.4.0
+		 *
+		 * @param string     $entity_type The type of the entity.
+		 * @param string|int $entity_id   The ID of the entity.
+		 */
+		do_action( 'woocommerce_entity_version_cache_deleted', $entity_type, $entity_id );
+
+		return $result;
+	}
+
+	/**
+	 * Validate entity type and entity ID inputs.
+	 *
+	 * @param string     $entity_type Entity type.
+	 * @param string|int $entity_id   Entity ID.
+	 * @return void
+	 * @throws \InvalidArgumentException If entity_type or entity_id are invalid.
+	 */
+	private function validate_input( string $entity_type, $entity_id ): void {
```

Comment:
Is this validation helpful beyond the empty string check?  It seems like we could get this for free by having stronger typing on our function arguments.